app/src/lib/shell.ts

```diff
@@ -0,0 +1,155 @@
+/* tslint:disable:no-sync-functions */
+
+import * as ChildProcess from 'child_process'
+import * as os from 'os'
+
+const ENVIRONMENT_VARIABLES_TO_PRESERVE = new Set([ 'NODE_ENV', 'NODE_PATH' ])
+
+type IndexLookup = {
+  [propName: string]: string;
+}
+
+/**
+ * Inspect whether the current process needs to be patched to get important
+ * environment variables for Desktop to work and integrate with other tools
+ * the user may invoke as part of their workflow.
+ *
+ * This is only applied to macOS installations due to how the application
+ * is launched.
+ *
+ * @param process The process to inspect.
+ */
+export function shellNeedsPatching(process: NodeJS.Process): boolean {
+  return __DARWIN__ && !process.env.PWD
+}
+
+type ShellResult = {
+  stdout: string,
+  error: Error | null,
+}
+
+/**
+ * Gets a dump of the user's configured shell environment.
+ *
+ * @returns the output of the `env` command or `null` if there was an error.
+ */
+async function getRawShellEnv(): Promise<string | null> {
+  const shell = getUserShell()
+
+  const promise = new Promise<ShellResult>((resolve) => {
+    let child: ChildProcess.ChildProcess | null = null
+    let error: Error | null = null
+    let stdout = ''
+    let done = false
+
+    // ensure we clean up eventually, in case things go bad
+    const cleanup = () => {
+      if (!done && child) {
+        child.kill()
+        done = true
+      }
+    }
+    process.once('exit', cleanup)
+    setTimeout(() => {
+      cleanup()
+    }, 5000)
+
+    const options =  { detached: true, stdio: [ 'ignore', 'pipe', process.stderr ] }
+
+    child = ChildProcess.spawn(shell, [ '-ilc', 'command env' ], options)
```

Comment:
Wasn't sure this will work for `csh` and family, because on my machine this happens:

```
$ /bin/csh --version
tcsh 6.18.01 (Astron) 2012-02-14 (x86_64-apple-darwin) options wide,nls,dl,bye,al,kan,sm,rh,color,filec
$ /bin/csh -ilc 'command env'
Unknown option: `-lc'
Usage: csh [ -bcdefilmnqstvVxX ] [ argument ... ].
```

But I can't find any explicit checks for this in `atom/atom` so 🤷‍♂️